src/util/include/util/RAII_Wrappers.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Xournal++
+ *
+ * RAII wrappers for C library classes
+ *
+ * @author Xournal++ Team
+ * https://github.com/xournalpp/xournalpp
+ *
+ * @license GNU GPLv2 or later
+ */
+
+#pragma once
+
+#include <cairo.h>
+
+namespace xoj::util {
+/**
+ * @brief Simple template class for RAII wrappers of Cairo/GTK classes
+ * @param T The wrapper will store a pointer of type T
+ * @param ref A function that increases the ref-count
+ * @param unref A function that decreases the ref-count
+ *
+ * WARNING: For GTK classes that often come as a floating ref.
+ *      e.g. https://docs.gtk.org/gobject/class.Object.html
+ *    call ref_sink before passing it to RAII_Wrapper(T*) or reset(T*).
+ */
+template <typename T, T* ref(T*), void unref(T*)>
+class RAII_Wrapper;
+
+/** SPECIALIZATIONS **/
+
+/* cairo */
+using CairoWrapper = RAII_Wrapper<cairo_t, cairo_reference, cairo_destroy>;
+using CairoSurfaceWrapper = RAII_Wrapper<cairo_surface_t, cairo_surface_reference, cairo_surface_destroy>;
+
+/**
+ * @brief cairo_save(cr)/cairo_restore(cr) RAII implementation
+ */
+class CairoSaveGuard;
+
+// TODO Add and use GTK specializations
+
+/********************/
+
+
+/** IMPLEMENTATION **/
+template <typename T, T* ref(T*), void unref(T*)>
+class RAII_Wrapper {
+public:
+    RAII_Wrapper() = default;
+    ~RAII_Wrapper() {
+        if (p) {
+            unref(p);
+        }
+    }
+
+    RAII_Wrapper(const RAII_Wrapper& other) {
+        if (other.p) {
+            p = ref(other.p);
+        }
+    }
+    RAII_Wrapper(RAII_Wrapper&& other) {
+        p = other.p;
+        other.p = nullptr;
+    }
+    RAII_Wrapper& operator=(const RAII_Wrapper& other) {
+        if (other.p) {  // if *this == &other, the ref count never drops to 0
+            ref(other.p);
+        }
+        if (p) {
+            unref(p);
+        }
+        p = other.p;
+        return *this;
+    }
+    RAII_Wrapper& operator=(RAII_Wrapper&& other) {
+        if (this == &other) {
+            return *this;
+        }
+        if (p) {
+            unref(p);
+        }
+        p = other.p;
+        other.p = nullptr;
+        return *this;
+    }
+
+    /**
+     * Assumes ownership is already established
+     *
+     * WARNING: This is not fit for GTK classes that come as a floating ref.
+     *      e.g. https://docs.gtk.org/gobject/class.Object.html
+     *    For such a pointer, call ref_sink before passing it to RAII_Wrapper(T*) or reset(T*).
+     */
+    RAII_Wrapper(T* p): p(p) {}
```

Comment:
I used this idea. I am not sure it is actually safer, unless we remove the default `Adopt = adopt`. Maybe the default should depend on the existence of H::ref_sink?

Reply:
In most cases, the programmer want to reference the pointer. 
If a widget (GInitiallyUnowned) is created, it's often passed to the future parent after the initialization. After that, the pointer has to be referenced anyway. 

So probably I would make Ref the default one.
Making it dependent may be counterintuitive, then we have a container with the same name but different behaviour. 
Smells like std::vector<bool>.

Also it would be cool, if you could add constexpr and noexcept were ever possible.

Reply:
> In most cases, the programmer want to reference the pointer.

This is just not true for Cairo objects (or even Poppler objects), for which the user almost never wants to add a reference.

Also, even for GWidget's, it is not clear that you would pass the widget first to its parent, and then to the smart pointer. If the user first creates the smart pointer, initializes a couple of parameters of the widget, and then adds it to the parent, then we need to ref_sink.
All in all, is the ref has already been sunk (by the parent, or anyone else), ref_sink is equivalent to ref and is still safe. I do not see any case where defaulting to ref could be better than defaulting to ref_sink (if ref_sink exists).

I think the bottom line is: we (I) want `SmartPointer p(create_object(...));` to be leak free: i.e. ref_sink for GObjects, and adopt for cairo/poppler things.

Reply:
Mhh, and when we add a type to H:

`using Default = Adopt` for example?
Then the constructor and the reset method must be templated against.